transform/utility_metric_freshness.go

```diff
@@ -0,0 +1,124 @@
+package transform
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"sync/atomic"
+	"time"
+
+	"github.com/brexhq/substation/config"
+	iconfig "github.com/brexhq/substation/internal/config"
+	"github.com/brexhq/substation/internal/errors"
+	"github.com/brexhq/substation/internal/metrics"
+	"github.com/brexhq/substation/message"
+)
+
+type utilityMetricFreshnessConfig struct {
+	Threshold string `json:"threshold"`
+
+	Object iconfig.Object `json:"object"`
+	Metric iconfig.Metric `json:"metric"`
+}
+
+func (c *utilityMetricFreshnessConfig) Decode(in interface{}) error {
+	return iconfig.Decode(in, c)
+}
+
+func (c *utilityMetricFreshnessConfig) Validate() error {
+	if c.Threshold == "" {
+		return fmt.Errorf("threshold: %v", errors.ErrMissingRequiredOption)
+	}
+
+	if c.Object.SourceKey == "" {
+		return fmt.Errorf("object_source_key: %v", errors.ErrMissingRequiredOption)
+	}
+
+	return nil
+}
+
+func newUtilityMetricFreshness(ctx context.Context, cfg config.Config) (*utilityMetricFreshness, error) {
+	conf := utilityMetricFreshnessConfig{}
+	if err := conf.Decode(cfg.Settings); err != nil {
+		return nil, fmt.Errorf("transform: utility_metric_freshness: %v", err)
+	}
+
+	if err := conf.Validate(); err != nil {
+		return nil, fmt.Errorf("transform: utility_metric_freshness: %v", err)
+	}
+
+	m, err := metrics.New(ctx, conf.Metric.Destination)
+	if err != nil {
+		return nil, fmt.Errorf("transform: utility_metric_freshness: %v", err)
+	}
+
+	dur, err := time.ParseDuration(conf.Threshold)
+	if err != nil {
+		return nil, fmt.Errorf("transform: utility_metric_freshness: duration: %v", err)
+	}
+
+	tf := utilityMetricFreshness{
+		conf:   conf,
+		metric: m,
+		dur:    dur,
+	}
+
+	return &tf, nil
+}
+
+type utilityMetricFreshness struct {
+	conf   utilityMetricFreshnessConfig
+	metric metrics.Generator
+	dur    time.Duration
+
+	success uint32
+	failure uint32
+}
+
+func (tf *utilityMetricFreshness) Transform(ctx context.Context, msg *message.Message) ([]*message.Message, error) {
+	// ctrl messages are handled by only one process, so the map
```

Comment:
docs: Should we replace the word "process" here with "thread" or "goroutine" since the concurrency primitive used in Substation isn't multi-processing?

Reply:
Yeah `thread` would be more accurate.